nodestream/pipeline/extractors/stores/aws/dynamodb_extractor.py

```diff
@@ -0,0 +1,199 @@
+from logging import getLogger
+from typing import Any, Dict, List, AsyncGenerator
+
+from ...credential_utils import AwsClientFactory
+from ...extractor import Extractor
+
+
+class DynamoObject:
+    """
+        Dynamo objects generally come in the form {"attribute_type": value}
+        Example:
+            {"S", "some_string"}
+            {"N", "12345"}
+    """
+    def __init__(self, object: Dict[str, Any]):
+        self.attribute_type = list(object.keys())[0]
+        self.value = list(object.values())[0]
+
+    def items(self) -> tuple[str, Any]:
+        return self.attribute_type, self.value
+    
+    def as_dynamo_representation(self):
+        if isinstance(self.value, dict):
+            return {self.attribute_type: {key: value.as_dynamo_representation() for key, value in self.value.items()}}
+        elif isinstance(self.value, list):
+            return {self.attribute_type: [value.as_dynamo_representation() for value in self.value ]}
+        return {self.attribute_type: self.value}
+
+def leave_untouched(value):
+    return value
+
+def convert_scalar(value, type_handler):
+    return  None if type_handler is None else type_handler(value)
+
+def convert_array(array: List[Any], type_handler):
+    return [type_handler(value) for value in array]
+
+def convert_map(mapping: Dict[str, Any], type_handler):
+    return {key: type_handler(value) for key, value in mapping.items()}
+
+def convert_number(value: str):
+    if '.' in value:
+        return float(value)
+    return int(value)
+
+DEFAULT_HANDLER = (convert_scalar, leave_untouched)
+
+def convert_value(dynamo_object: Dict[str, Any]):
+    attribute_type, value = DynamoObject(dynamo_object).items()
+    result_handler, type_handler = ATTRIBUTE_TYPE_DESCRIPTOR_HANDLERS.get(attribute_type, DEFAULT_HANDLER)
+    return result_handler(value, type_handler)
+
+
+ATTRIBUTE_TYPE_DESCRIPTOR_HANDLERS = {
+    "S": (convert_scalar, str),
+    "N": (convert_scalar, convert_number),
+    "B": (convert_scalar, bytes),
+    "NULL": (convert_scalar, None),
+    "BOOL": (convert_scalar, bool),
+    "SS": (convert_array, str),
+    "NS": (convert_array, convert_number),
+    "BS": (convert_array, bytes),
+    "M": (convert_map, convert_value),
+    "L": (convert_array, convert_value)
+}
+
+class DynamoRowConverter:
+    def __init__(self, item: Dict[str, Any]) -> None:
+        self.item = item
+
+    # { key: {"S": "string"}, secondary_key: { "M": {Stuff} }, tertiary_key: { "L": [Stuff] } }
+    def convert_row(self):
+        return {attribute_name: convert_value(attribute_value) for attribute_name, attribute_value in self.item.items()}
+
+
+def verify_list_consistency(value_list: List[Any]):
+    type_set = set([type(value) for value in value_list])
+    if len(type_set > 1):
+        return "L"
+    return REVERSE_ORDER_DESCRIPTOR_HANDLERS[type_set[0]] + "S"
```

Comment:
This is clever in the generic, but it doesn't work for all inputs. https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html

For instance, a list of maps needs to stay as a list of maps because there is no map set. 

I'd check if its part of the 'set' types.